File: src/ecs/systems/fleet.ts
import {
  DEFAULT_DRONE_BATTERY,
  DEFAULT_DRONE_CAPACITY,
  DEFAULT_DRONE_MINING_RATE,
  DEFAULT_DRONE_SPEED,
  type DroneEntity,
  type GameWorld,
  removeDrone,
  spawnDrone,
} from '@/ecs/world';
import { RESOURCE_KEYS } from '@/lib/biomes';
import { getResourceModifiers, type ResourceModifierSnapshot } from '@/lib/resourceModifiers';
import type { Modules, StoreApiType } from '@/state/store';

const updateDroneStats = (
  drone: DroneEntity,
  modules: Modules,
  modifiers: ResourceModifierSnapshot,
) => {
  const speedBonus = 1 + Math.max(0, modules.droneBay - 1) * 0.05;
  const baseSpeed = DEFAULT_DRONE_SPEED * speedBonus;
  drone.speed = baseSpeed * modifiers.droneProductionSpeedMultiplier;

  const capacityBase = DEFAULT_DRONE_CAPACITY + modules.storage * 5;
  const targetCapacity = capacityBase * modifiers.droneCapacityMultiplier;
  if (drone.cargo > targetCapacity) {
    const scale = targetCapacity > 0 ? targetCapacity / drone.cargo : 0;
    if (scale <= 0) {
      drone.cargo = 0;
      for (const key of RESOURCE_KEYS) {
        drone.cargoProfile[key] = 0;
      }
    } else {
      for (const key of RESOURCE_KEYS) {
        drone.cargoProfile[key] *= scale;
      }
      drone.cargo = targetCapacity;
    }
  }
  drone.capacity = targetCapacity;

  const miningBase = DEFAULT_DRONE_MINING_RATE + modules.refinery * 0.5;
  drone.miningRate = miningBase * modifiers.droneProductionSpeedMultiplier;

  const previousMax = drone.maxBattery > 0 ? drone.maxBattery : DEFAULT_DRONE_BATTERY;
  const fraction = previousMax > 0 ? drone.battery / previousMax : 0;
  const targetMaxBattery = DEFAULT_DRONE_BATTERY * modifiers.droneBatteryMultiplier;
  drone.maxBattery = targetMaxBattery;
  drone.battery = Math.min(drone.maxBattery, Math.max(0, fraction * drone.maxBattery));
};

export const createFleetSystem = (world: GameWorld, store: StoreApiType) => {
  const { droneQuery, factory } = world;
  const removeDroneReferences = (droneId: string) => {
    const state = store.getState();
    for (const candidate of state.factories) {
      if (candidate.queuedDrones.includes(droneId)) {
        state.undockDroneFromFactory(candidate.id, droneId);
      }
    }
  };

  return (_dt: number) => {
    const { modules, resources } = store.getState();
    const modifiers = getResourceModifiers(resources);
    const target = Math.max(1, modules.droneBay);
    while (droneQuery.size < target) {
      const drone = spawnDrone(world);
      const primaryFactory = store.getState().factories[0];
      drone.position.copy(primaryFactory?.position ?? factory.position);
    }
    while (droneQuery.size > target) {
      const drone = droneQuery.entities[droneQuery.size - 1];
      removeDroneReferences(drone.id);
      removeDrone(world, drone);
    }
    for (const drone of droneQuery) {
      updateDroneStats(drone, modules, modifiers);
    }
  };
};

File: rust-engine/src/systems/fleet.rs
use crate::constants::{
    DRONE_MAX_BATTERY, DRONE_MAX_CARGO, DRONE_MINING_RATE, DRONE_SPEED,
};
use crate::modifiers::ResourceModifierSnapshot;
use crate::schema::Modules;

pub fn sys_fleet(
    battery: &mut [f32],
    max_battery: &mut [f32],
    capacity: &mut [f32],
    mining_rate: &mut [f32],
    cargo: &mut [f32],
    cargo_profile: &mut [f32], // 5 floats per drone
    modules: &Modules,
    modifiers: &ResourceModifierSnapshot,
) {
    let drone_count = battery.len();

    // Calculate base stats
    let speed_bonus = 1.0 + (modules.drone_bay as f32 - 1.0).max(0.0) * 0.05;
    let _base_speed = DRONE_SPEED * speed_bonus;
    // let target_speed = base_speed * modifiers.drone_production_speed_multiplier;

    let capacity_base = DRONE_MAX_CARGO + (modules.storage as f32) * 5.0;
    let target_capacity = capacity_base * modifiers.drone_capacity_multiplier;

    let mining_base = DRONE_MINING_RATE + (modules.refinery as f32) * 0.5;
    let target_mining_rate = mining_base * modifiers.drone_production_speed_multiplier;

    let target_max_battery = DRONE_MAX_BATTERY * modifiers.drone_battery_multiplier;

    for i in 0..drone_count {
        // Update capacity
        if cargo[i] > target_capacity {
            let scale = if target_capacity > 0.0 { target_capacity / cargo[i] } else { 0.0 };
            if scale <= 0.0 {
                cargo[i] = 0.0;
                for j in 0..5 {
                    cargo_profile[i * 5 + j] = 0.0;
                }
            } else {
                cargo[i] = target_capacity;
                for j in 0..5 {
                    cargo_profile[i * 5 + j] *= scale;
                }
            }
        }
        capacity[i] = target_capacity;

        // Update mining rate
        mining_rate[i] = target_mining_rate;

        // Update battery
        let was_uninitialized = max_battery[i] == 0.0;
        let previous_max = if !was_uninitialized { max_battery[i] } else { DRONE_MAX_BATTERY };
        let fraction = if previous_max > 0.0 { battery[i] / previous_max } else { 0.0 };

        max_battery[i] = target_max_battery;

        if was_uninitialized {
            battery[i] = target_max_battery;
        } else {
            // Scale battery to new max, preserving percentage
            battery[i] = (fraction * target_max_battery).min(target_max_battery).max(0.0);
        }
    }
}
